core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedStreamSinkChannel.java

```diff
@@ -527,12 +527,16 @@ public boolean isOpen() {
     }
 
     @Override
-    public synchronized void close() throws IOException {
```

Comment:
@fl4via I didn't see anywhere else that synchronizes on the instance so I assumed that this was added to prevent multiple invocations of the close method body, but I could be mistaken.

Reply:
@carterkozak it was make synchronized to protect from the removal of synchronization from AbstractFramedChannel.receive()
That said, maybe this fix also works. Let's see first what is made of https://issues.redhat.com/browse/WFLY-12822, to see if the previous fix of UNDERTOW-1643 will be kept

Reply:
Thanks for the information!

If #852 merges, I think we'll still want to keep the changes to check STATE_CLOSED under the lock in order to avoid multiple concurrent executions.

Tagging @ropalka for context as well

Reply:
We must avoid as much locks as possible. Introducing synchronization on AbstractFramedStreamSinkChannel object itself was not ideal step forward to fix deadlock IMHO.

Reply:
Yes, I completely agree, I'm referring to the existing locking in the close method on line 534. On master two threads can call close at the same time, both see that the channel is not closed, and each lock before setting `state |= STATE_CLOSED;`. Without either a check under synchronization or a compare-and-swap check, it will still be possible to run the body of the close method twice, which we expect to be run once.

Reply:
@carterkozak I agree with you on that. We need to fix that as well.